File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans= new ArrayList<>();
        backtrack(0, nums, new ArrayList<>(), ans);
        return ans;
    }
    public void backtrack(int idx, int[]nums, List<Integer> cur, List<List<Integer>> ans){
        ans.add(new ArrayList<>(cur));
        for(int i=idx; i<nums.length;i++){
            cur.add(nums[i]);
            backtrack(i+1, nums, cur, ans);
            cur.remove(cur.size()-1);
        }
    }
}

File: 78-subsets/subsets.cpp
class Solution {
public:
    void set(vector<int> &temp,vector<int> & nums, int st ,int end,vector<vector<int>> &arr){
        if(st>end){
            arr.push_back(temp);
            return;
        }
        // not pick
        set(temp,nums,st+1,end,arr);
        // pick
        temp.push_back(nums[st]);
        set(temp,nums,st+1,end,arr);
        temp.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> arr;
        vector<int> temp;
        set(temp,nums,0,nums.size()-1,arr);
        return arr;
    }
};

